qemu/tests/pci_bridge.py

```diff
@@ -0,0 +1,286 @@
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    [pci-bridge] Check devices which attached to 1 pci-bridge
+    [pci-bridge] Check devices which attached to different pci-bridges
+    [pci-bridge] Hotplug and unplug different devices to 1 pci-bridge
+    [pci-bridge] Hotplug and unplug different devices to different pci-bridges
+    This case will:
+    1) Attach one or multiple pci-bridge(s) to guest.
+    2) Prepare devices attached to pci-bridge(s).
+    3) Start the guest.
+    4) Hotplug devices (if needed).
+    5) Check devices.
+    6) Hotunplug devices (if needed).
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    error_context.context("Modify params!", logging.info)
+    pci_bridge_num = int(params.get("pci_bridge_num", 1))
+    prepare_pci_bridge(test, params, pci_bridge_num)
+    pci_bridges = params.objects("pci_controllers")
+
+    device_num = 0
+    image_num = int(params.get("image_num", 3))
+    device_num = prepare_images(test, params, image_num, device_num)
+
+    if params.get("need_hotplug_unplug", "no") == "no":
+        params["nics"] = "nic0"
+        params["nic_model_nic0"] = "virtio"
+        params["device_id_nic0"] = "nic0"
+        if pci_bridge_num == 1:
+            if device_num >= 31:
+                test.fail("There are already %d devices on %s"
+                          % (device_num, pci_bridges[0]))
+            params["nic_pci_bus_nic0"] = pci_bridges[0]
+        else:
+            index = device_num % pci_bridge_num
+            params["nic_pci_bus_nic0"] = pci_bridges[index]
+            device_num += 1
+
+    params["start_vm"] = "yes"
+    env_process.process_images(env_process.preprocess_image, test, params)
+    env_process.preprocess_vm(test, params, env, params["main_vm"])
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    login_timeout = int(params.get("login_timeout", 360))
+    session = vm.wait_for_login(timeout=login_timeout)
+
+    if params.get("need_hotplug_unplug", "no") == "yes":
+        device_list = []
+        fmt_list = params.objects("disk_driver")
+        image_list = params.objects("images")
+        del image_list[0]
+        for i in range(image_num):
+            if i >= len(fmt_list):
+                test.error("No sufficient disk driver type!")
+            if pci_bridge_num == 1:
+                pci_bridge_id = pci_bridges[0]
+            else:
+                index = i % pci_bridge_num
+                pci_bridge_id = pci_bridges[index]
+            error_context.context("Hotplug a %s disk on %s!"
+                                  % (fmt_list[i], pci_bridge_id), logging.info)
+            device_list += disk_hotplug(test, params, vm, session,
+                                        image_list[i], fmt_list[i], pci_bridge_id)
+
+    check_data_disks(test, params, env, vm, session)
```

Comment:
Please close the session when not use it anymore.

Reply:
Modified, Thanks!